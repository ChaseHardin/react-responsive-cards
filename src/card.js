import React from 'react'
import Card from 'react-bootstrap/Card'

export default ({ detail }) => {
  const makeAltText = `Picture for ${detail.title}`
  const renderImage = () =>
    detail.image && (
      <Card.Img
        alt={makeAltText}
        variant='top'
        src={detail.image}
        style={{ width: '100%', height: '400px', objectFit: 'cover' }}
      />
    )

  return (
    <Card>
      {renderImage()}
      <Card.Body>
        <Card.Title>{detail.title}</Card.Title>
        <Card.Text>{detail.description}</Card.Text>
      </Card.Body>
    </Card>
  )
}
