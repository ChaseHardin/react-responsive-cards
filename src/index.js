import React from 'react'
import 'bootstrap/dist/css/bootstrap.min.css'
import Card from './card'
import Row from 'react-bootstrap/Row'
import Col from 'react-bootstrap/Col'
import Container from 'react-bootstrap/Container'

export const Cards = (props) => {
  const renderCards = () =>
    props.details.map((detail, index) => (
      <Col
        key={index}
        sm={12}
        md={6}
        lg={4}
        style={{ marginBottom: '1rem', marginTop: '2rem' }}
      >
        <Card detail={detail} />
      </Col>
    ))

  return (
    <Container>
      <Row>{renderCards()}</Row>
    </Container>
  )
}
