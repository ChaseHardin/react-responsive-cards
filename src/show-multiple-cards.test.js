import React from 'react'
import { render } from '@testing-library/react'
import { Cards } from '../src/index'
import chance from '../src/test-utils/chance-util'
import '@testing-library/jest-dom/extend-expect'
import renderer from 'react-test-renderer';

test('should allow a user to view multiple cards', () => {
  const details = [
    {
      title: chance.string(),
      description: chance.paragraph(),
      image: chance.url()
    },
    {
      title: chance.string(),
      description: chance.paragraph(),
      image: chance.url()
    }
  ]

  const { getByText, getByAltText, container } = render(
    <Cards details={details} />
  )

  details.forEach((detail) => {
    getByText(detail.title)
    getByText(detail.description)
    getByAltText(`Picture for ${detail.title}`)
    expect(getByAltText(`Picture for ${detail.title}`))
      .toHaveAttribute(
        'src',
        detail.image
      )
    expect(getByAltText(`Picture for ${details[0].title}`))
      .toHaveAttribute(
        'style',
        'width: 100%; height: 20rem; object-fit: cover;'
      )
    expect(container.querySelector(`[data-card="${details[0].title}"]`))
      .toHaveAttribute(
        'style',
        'box-shadow: 0 4px 8px 0 rgba(0, 0, 0, 0.2), 0 6px 20px 0 rgba(0, 0, 0, 0.19);'
      )

    expect(container.querySelector(`[data-card-text="${details[0].title}"]`))
      .toHaveAttribute(
        'style',
        'height: 8rem; overflow: scroll;'
      )
  })
})

test('should allow a user to provide no images', () => {
  const details = [
    {
      title: chance.string(),
      description: chance.paragraph()
    }
  ]

  const { getByAltText } = render(<Cards details={details} />)

  expect(getByAltText(`Picture for ${details[0].title}`)).toHaveAttribute(
    'src',
    'https://via.placeholder.com/500x500.png?text=react-responsive-cards'
  )
})

it('renders cards with correct bootstrap styles', () => {
  const details = [{
    title: 'The Card Title',
    description: 'This is a short description',
    image: 'https://<image_here>.jpg',
    variant: 'dark',
    buttonText: 'Details',
    callbackFn: () => {}
  }]

  const tree = renderer
    .create(<Cards details={details}/>)
    .toJSON();

  expect(tree).toMatchSnapshot();
});
