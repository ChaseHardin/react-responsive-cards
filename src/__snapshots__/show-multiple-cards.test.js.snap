// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders cards with correct material-ui styles 1`] = `
<div
  className="MuiGrid-root MuiGrid-container MuiGrid-justify-xs-center"
>
  <div
    className="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-align-items-xs-center MuiGrid-justify-xs-center MuiGrid-grid-sm-6 MuiGrid-grid-md-4 MuiGrid-grid-lg-3"
  >
    <div
      className="MuiPaper-root MuiCard-root makeStyles-root-5 MuiPaper-elevation1 MuiPaper-rounded"
      style={
        Object {
          "margin": "6px",
        }
      }
    >
      <button
        className="MuiButtonBase-root MuiCardActionArea-root"
        disabled={false}
        onBlur={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <div
          className="MuiCardMedia-root makeStyles-media-6"
          data-card-media="The Card Title"
          style={
            Object {
              "backgroundImage": "url(\\"https://<image_here>.jpg\\")",
            }
          }
          title="The Card Title"
        />
        <div
          className="MuiCardContent-root"
        >
          <h2
            className="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom"
          >
            The Card Title
          </h2>
          <p
            className="MuiTypography-root MuiTypography-body2 MuiTypography-colorTextSecondary"
          >
            This is a short description
          </p>
        </div>
        <span
          className="MuiCardActionArea-focusHighlight"
        />
      </button>
    </div>
  </div>
</div>
`;
